### DISCUSSION TITLE
integrating with pir motion sensor

### DISCUSSION THREAD

**Discussion by Manish-0521** [Geektoolkit/Dynaframe3#91] [Q&A]
Can anyone please let me know how to integrate Dynaframe with pir motion sensor. I know a bit about python, can anyone send the link to the code for Dynaframe. So that i can make changes in the code to run it with a PIR sensor.

**Comment by RichN001** [comment]
hey, you just need to do this on the Pi side, Dynaframe does not need to be altered for this. I have a setup for this that I use. 

If you have a sensor connect it to your Pi on GPIO 18

On your Pi with Dynaframe either ssh or on console

`sudo wget http://xgraphics.com/motion/install.sh && bash install.sh; rm -f install.sh; sudo reboot`

This will install my script in /home/pi/motion and reboot your frame, So it will setup the motion script to run on boot. If you do not want it to auto start after install rm the following. 

`rm -rf /home/pi/motion
rm -rf /home/pi/.config/autostart/motion.desktop`


If your sensor is connected properly with no motion the screen will turn off after 30 seconds. You can change this Delay in /home/pi/motion/motion.py on line 15. Its in seconds. (I find 60mins (3600) is enough in my living space that's active)

What the script does... 

Lets the sensor rest before starting, then looks for motion. When it sees motion it starts the delay time... if it sees motion again it resets the delay time. when delay time is reached, the monitor will go off and obviously turns on when it first sees motion. 

Im hoping to get this in to being a part of Dynaframe with options for settings in the WebApp. But Ui is my priority when I get back to working on things. I made the install to make it easy for me to run on new frames.. enjoy.

**Reply by Manish-0521** [reply]
Thank you soo much for the help. :)

**Comment by RichN001** [comment]
here is the script if anyone just wants that and doesn't want to use the install linked above. Save below to filename.py chmod a+x filename.py and python3 filename.py to run it. 

[motion.py]( http://xgraphics.com/motion/motion.py)

**Reply by RichN001** [reply]
@Manish-0521 Let me know if you get it working ok..

**Reply by Manish-0521** [reply]
yupp it working fine. thanks for the fast reply. Actually i want to know if you can stop and restart the actual dynaframe loop pictures while there is no motion. Cause i guess the Dynaframe is running in the background. Actually what im trying to do is to start the loop from the beginning when the motion is detected, cause i want to great someone when they enter the room with a welcome sign and with sound. Let me know if you have any suggestion and Thank for the help :) .

**Reply by RichN001** [reply]
if you name your photos or video accordingly.. 123.., etc... you could probably run a http post request to dynaframe with slideshow first.. to start on wake up. i'll have to look for the api.. should work thou

**Reply by RichN001** [reply]
yourframeip:8000/default.htm?COMMAND=CONTROL_FIRST

**Reply by Manish-0521** [reply]
okkk Thanks :)

**Reply by RichN001** [reply]
I don't have a sensor here to test, but should work...

add new line `import request` on the line after `import subprocess`

then new line after `subprocess.call('vcgencmd display_power 1',shell=True)`  with
`url = 'http://yourip:8000/default.htm?COMMAND=CONTROL_FIRST'`

That should start your slideshow with the first item on wake up, depending on your slide duration you may want to pause the url to give time for the screen to wake.

**Reply by Manish-0521** [reply]
I have tried this just now, its not working.

**Reply by RichN001** [reply]
sorry import request should be import requests with a s...

**Reply by Manish-0521** [reply]
i have found a solution to this. What i did is 
declared movie_path = '/path/to/your/file.mpg' at the begining, then 

after the line subprocess.call('vcgencmd display_power 1',shell=True), i added a new line 
from subprocess import Popen
omxp = Popen(['omxplayer',movie_path]).
this worked perfectly try for me to use this, i used this code just to play a welcome video when motion is detedcted.
Thank you for the pir motion code , it was a great help :)...

**Reply by RichN001** [reply]
right on, glad you got it working... I have it working with images.. didn't try movie.

**Comment by RichN001** [comment]
Just wanted to follow up on this thread in case anyone else wants to do this. Here is updated Python code, this uses the url API properly. This will reflect the button for Turn Screen on and off as it writes the screen status to the app settings. With testing the above call CONTROL_FIRST I found it was not actually working, well it wasn't actually requesting the URL. When I tested it I just did it in its own script I must of forgotten something when I put it in the original motion code. I didn't test this either for multiple commands but I can't imagine it won't work if you just copy this block and put in different command. Also can just use localhost this way it will always work. 

```
        # Turns Dynaframe Screen On
        link = "http://localhost:8000/default.htm?COMMAND=SCREENON"
        f = requests.get(link)
        # Copy above section to add another dynaframe URL
```
  

```
#!/usr/bin/python3
import RPi.GPIO as GPIO
import time
import os
import subprocess
import requests


GPIO.setmode(GPIO.BCM)

# Define GPIO to use on Pi
GPIO_PIR = 24
print ("PIR Module Test (CTRL-C to exit)")

GPIO.setup(GPIO_PIR,GPIO.IN)      # Echo

delay = 30
try:

    print ("Waiting for PIR to settle ...")

    # Loop until PIR output is 0
    while GPIO.input(GPIO_PIR)==1:
        time.sleep(10)

    print ("Ready")

    # Loop until users quits with CTRL-C
    while True :

        # Read PIR state
        while GPIO.input(GPIO_PIR) == 0:
            time.sleep(0.1)
            
        start = time.time()
        # PIR is triggered
        print ("Motion detected!")       
        time.sleep(0.5) 
        # Turns Dynaframe Screen On
        link = "http://localhost:8000/default.htm?COMMAND=SCREENON"
        f = requests.get(link)
        # Copy above section to add another dynaframe URL
  

        # start timer  with reset if pir retriggered    
        end = start + delay
        now = time.time()
        while end > now:  
            if GPIO.input(GPIO_PIR) == 1:
                print ("Motion detected again!")
                start = time.time()
                end = start + delay
            now = time.time()
            time.sleep(0.1)
                
        print ("Time up")
        
        # Turns Dynaframe Screen Off
        link = "http://localhost:8000/default.htm?COMMAND=SCREENOFF"
        f = requests.get(link)
        # Copy above section to add another dynaframe URL
  


except KeyboardInterrupt:
  print ("  Quit")
  # Reset GPIO settings
  GPIO.cleanup()

```
